src/auth_context.rs

```diff
@@ -0,0 +1,155 @@
+    /// considered valid by gRPC).
+    /// `false` in non-secure scenarios.
+    pub fn peer_is_authenticated(&self) -> bool {
+        assert!(!self.ctx.is_null());
+        unsafe { grpc_sys::grpc_auth_context_peer_is_authenticated(self.ctx) != 0 }
+    }
+
+    /// `AuthContext[peer_identity_property_name()]`
+    ///
+    /// There may be several of them (for instance if `x509_subject_alternative_name` is selected)
+    pub fn peer_identity(&self) -> AuthPropertyIter {
+        assert!(!self.ctx.is_null());
+        unsafe {
+            // grpc_auth_context_peer_identity returns empty_iterator when self.ctx is NULL
+            let iter = grpc_sys::grpc_auth_context_peer_identity(self.ctx);
+            AuthPropertyIter {
+                iter,
+                _lifetime: PhantomData,
+            }
+        }
+    }
+}
+
+impl<'a> IntoIterator for &'a AuthContext {
+    type Item = AuthProperty<'a>;
+    type IntoIter = AuthPropertyIter<'a>;
+
+    /// Iterate over the AuthContext properties
+    fn into_iter(self) -> Self::IntoIter {
+        unsafe {
+            // grpc_auth_context_property_iterator returns empty_iterator when self.ctx is NULL
+            let iter = grpc_sys::grpc_auth_context_property_iterator(self.ctx);
+            AuthPropertyIter {
+                iter,
+                _lifetime: PhantomData,
+            }
+        }
+    }
+}
+
+impl Drop for AuthContext {
+    fn drop(&mut self) {
+        unsafe { grpc_sys::grpc_auth_context_release(self.ctx) }
+    }
+}
+
+pub struct AuthPropertyIter<'a> {
+    iter: grpc_auth_property_iterator,
+    _lifetime: PhantomData<&'a grpc_auth_property_iterator>,
+}
+
+impl<'a> Iterator for AuthPropertyIter<'a> {
+    type Item = AuthProperty<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // grpc_auth_property_iterator_next returns empty_iterator when self.iter is NULL
+        let prop = unsafe { grpc_sys::grpc_auth_property_iterator_next(&mut self.iter) };
+        if prop.is_null() {
+            None
+        } else {
+            Some(AuthProperty {
+                prop,
+                _lifetime: PhantomData,
+            })
+        }
+    }
+}
+
+/// Auth properties are elements of the AuthContext. They have a name
+/// (a key of type string) and a value which can be a string or binary data.
+pub struct AuthProperty<'a> {
+    prop: *const grpc_auth_property,
+    _lifetime: PhantomData<&'a grpc_auth_property>,
+}
+
+impl<'a> AuthProperty<'a> {
+    pub fn name(&self) -> &'a str {
+        unsafe {
+            CStr::from_ptr((*self.prop).name)
+                .to_str()
+                .expect("Auth property name should be valid UTF-8 data")
+        }
+    }
+
+    pub fn value(&self) -> &'a [u8] {
+        unsafe {
+            std::slice::from_raw_parts((*self.prop).value as *const u8, (*self.prop).value_length)
+        }
+    }
+
+    /// (name, value) tuple for easier pattern matching
+    pub fn pair(&self) -> (&'a str, &'a [u8]) {
```

Comment:
Let's remove `pair` and `str_pair` and provide a `value_str` method to keep minimal interfaces.